internal/manifestgen/manifestgen.go

```diff
@@ -0,0 +1,178 @@
+		if err != nil {
+			return nil, err
+		}
+		containerSpecs[plName] = specs
+	}
+	return containerSpecs, nil
+}
+
+func defaultCommitResolver(commitSources map[string][]ostree.SourceSpec) (map[string][]ostree.CommitSpec, error) {
+	commits := make(map[string][]ostree.CommitSpec, len(commitSources))
+	for name, commitSources := range commitSources {
+		commitSpecs := make([]ostree.CommitSpec, len(commitSources))
+		for idx, commitSource := range commitSources {
+			var err error
+			commitSpecs[idx], err = ostree.Resolve(commitSource)
+			if err != nil {
+				return nil, err
+			}
+		}
+		commits[name] = commitSpecs
+	}
+	return commits, nil
+}
+
+type (
+	DepsolveFunc func(cacheDir string, packageSets map[string][]rpmmd.PackageSet, d distro.Distro, arch string) (map[string][]rpmmd.PackageSpec, map[string][]rpmmd.RepoConfig, error)
+
+	ContainerResolverFunc func(containerSources map[string][]container.SourceSpec, archName string) (map[string][]container.Spec, error)
+
+	CommitResolverFunc func(commitSources map[string][]ostree.SourceSpec) (map[string][]ostree.CommitSpec, error)
+)
+
+// Options contains the optional settings for the manifest generation.
+// For unset values defaults will be used.
+type Options struct {
+	Cachedir          string
+	Output            io.Writer
+	Depsolver         DepsolveFunc
+	ContainerResolver ContainerResolverFunc
+	CommitResolver    CommitResolverFunc
+}
+
+// Generator can generate an osbuild manifest from a given repository
+// and options.
+type Generator struct {
+	cacheDir string
+	out      io.Writer
+
+	depsolver         DepsolveFunc
+	containerResolver ContainerResolverFunc
+	commitResolver    CommitResolverFunc
+
+	reporegistry *reporegistry.RepoRegistry
+}
+
+// New will create a new manifest generator
+func New(reporegistry *reporegistry.RepoRegistry, opts *Options) (*Generator, error) {
+	if opts == nil {
+		opts = &Options{}
+	}
+	mg := &Generator{
+		reporegistry: reporegistry,
+
+		cacheDir:          opts.Cachedir,
+		out:               opts.Output,
+		depsolver:         opts.Depsolver,
+		containerResolver: opts.ContainerResolver,
+		commitResolver:    opts.CommitResolver,
+	}
+	if mg.out == nil {
+		mg.out = os.Stdout
+	}
+	if mg.depsolver == nil {
+		mg.depsolver = defaultDepsolver
+	}
+	if mg.containerResolver == nil {
+		mg.containerResolver = defaultContainerResolver
+	}
+	if mg.commitResolver == nil {
+		mg.commitResolver = defaultCommitResolver
+	}
+
+	return mg, nil
+}
+
+// Generate will generate a new manifest for the given distro/imageType/arch
+// combination.
+func (mg *Generator) Generate(bp *blueprint.Blueprint, dist distro.Distro, imgType distro.ImageType, a distro.Arch, imgOpts *distro.ImageOptions) error {
+	if imgOpts == nil {
+		imgOpts = &distro.ImageOptions{}
+	}
+
+	repos, err := mg.reporegistry.ReposByImageTypeName(dist.Name(), a.Name(), imgType.Name())
+	if err != nil {
+		return err
+	}
+	preManifest, warnings, err := imgType.Manifest(bp, *imgOpts, repos, 0)
```

Comment:
What about the seed here? I think the we should make it random by default. A fixed seed should be opt-in.

Reply:
Thanks, (very) nice catch and *urg* for missing this. 

Thinking a bit about it I wonder^Wthink we should look into images and tweak the API here, it seems in almost all "real" use-cases one would want a non-zero seed here so maybe this could become either a pointer (and on nil we just pick a random seed) or keep as is but assign a random value if we get "0" in images. Obviously followup material and I fixed this instance of the issue here but thinking about the future/a followup, wdyt?